wincode/src/io.rs

```diff
@@ -220,184 +310,466 @@ impl<'a> Writer<'a> {
     #[inline]
-    pub unsafe fn write_slice_t<T>(&mut self, value: &[T]) -> WriteResult<()> {
+    unsafe fn write_from_slice_t<T>(&mut self, src: &[T]) -> WriteResult<()> {
+        let dst = self.get_slice_mut(size_of_val(src))?;
+        // SAFETY:
+        // - `get_slice_mut` must do the appropriate bounds checking.
         unsafe {
-            self.write_exact(slice::from_raw_parts(
-                value.as_ptr().cast(),
-                size_of_val(value),
-            ))
+            ptr::copy_nonoverlapping(src.as_ptr().cast(), dst.as_mut_ptr(), size_of_val(src))
+        };
+        Ok(())
+    }
+}
+
+/// In-memory [`Writer`] that does not perform bounds checking.
+///
+/// Methods will panic if writes go out of bounds, so ensure that
+/// the chain of [`SchemaWrite`](crate::SchemaWrite) implementations that
+/// follow have statically known size requirements.
+pub struct TrustedSliceWriter<'a> {
+    buffer: &'a mut [MaybeUninit<u8>],
+    pos: usize,
+    /// The original length of the source that created the trusted writer.
+    ///
+    /// This can be used to check for trailing bytes when the writer is finished.
+    source_len: usize,
+    /// The global position of the writer relative to the original source.
+    global_pos: usize,
+}
+
+impl<'a> TrustedSliceWriter<'a> {
+    /// Create a new [`TrustedSliceWriter`] from a fresh buffer.
+    ///
+    /// Prefer [`TrustedSliceWriter::from_source`] if you're starting from an untrusted buffer.
+    pub fn new(buffer: &'a mut [MaybeUninit<u8>]) -> Self {
+        Self::from_source(buffer, buffer.len(), 0)
+    }
+
+    /// Create a new [`TrustedSliceWriter`] from an untrusted buffer, providing
+    /// the length of the original source.
+    ///
+    /// Prefer [`TrustedSliceWriter::new`] if you're starting from a fresh buffer.
+    pub fn from_source(
+        buffer: &'a mut [MaybeUninit<u8>],
+        source_len: usize,
+        global_pos: usize,
+    ) -> Self {
+        Self {
+            buffer,
+            pos: 0,
+            source_len,
+            global_pos,
         }
     }
 }
 
+impl<'a> Writer for TrustedSliceWriter<'a> {
+    type Trusted<'b>
+        = TrustedSliceWriter<'b>
+    where
+        Self: 'b;
+
+    #[inline]
+    #[allow(clippy::arithmetic_side_effects)]
+    fn finish(self) -> usize {
+        self.global_pos + self.pos
+    }
+
+    #[inline]
+    #[allow(clippy::arithmetic_side_effects)]
+    fn finish_disallow_trailing_bytes(self) -> WriteResult<usize> {
+        let final_pos = self.global_pos + self.pos;
+        // Pos should never write beyond the source length.
+        let rem = self.source_len - final_pos;
+        if rem != 0 {
+            return Err(writer_trailing_bytes(rem));
+        }
+
+        Ok(final_pos)
+    }
+
+    #[inline]
+    #[allow(clippy::arithmetic_side_effects)]
+    fn get_slice_mut(&mut self, len: usize) -> WriteResult<&mut [MaybeUninit<u8>]> {
+        let dst = &mut self.buffer[self.pos..self.pos + len];
```

Comment:
here and in all the other trusted impls - can't you use get_unchecked if this is trusted?